### PR TITLE
[DPE-4106] Tests legacy and modern endpoints simultaneously

### PR DESCRIPTION
**Pull request by BalabaDmitri**
**Issue** #444 
Simultaneous creation of legacy and modern endpoints relations

**Solution**
Charm goes into a "blocked" state, when creating a simultaneous relation legacy and modern endpoints.

### CODE REVIEW

**Comment by taurus-forever** [comment]
Hi @BalabaDmintri  and thank you for contribution here!
Please sign contribution agreement to merge the PR: https://github.com/canonical/postgresql-operator/actions/runs/8256188989/job/22618992227?pr=396

Tl;DR: Are we testing the right things here?  :-D

Long: I am slightly confused. I thought our current position is to [disallow simultaneous legacy and modern interfaces](https://charmhub.io/postgresql/docs/e-legacy-charm): `Please choose one endpoint to use. No need to relate all of them simultaneously!`. 
IMHO, charm should get into 'blocked' state forcing users to remove one relation and choose legacy or modern only.
@marceloneppel is it still valid position? I cannot find such blockers in charm.py code :-(

**Comment by marceloneppel** [comment]
> Hi @BalabaDmintri and thank you for contribution here! Please sign contribution agreement to merge the PR: https://github.com/canonical/postgresql-operator/actions/runs/8256188989/job/22618992227?pr=396
> 
> Tl;DR: Are we testing the right things here? :-D
> 
> Long: I am slightly confused. I thought our current position is to [disallow simultaneous legacy and modern interfaces](https://charmhub.io/postgresql/docs/e-legacy-charm): `Please choose one endpoint to use. No need to relate all of them simultaneously!`. IMHO, charm should get into 'blocked' state forcing users to remove one relation and choose legacy or modern only. @marceloneppel is it still valid position? I cannot find such blockers in charm.py code :-(

Hi @taurus-forever! In my opinion, we need to fix this in the relations to forbid them from being established simultaneously (and match `Note: do NOT relate both modern and legacy interfaces simultaneously!` from https://charmhub.io/postgresql/docs/e-interfaces). We should allow it only if we find a valid use case.

If we agree on that, I can create issues for that.

**Review by taurus-forever — CHANGES_REQUESTED** [review]
@BalabaDmintri , can you please add the test and charm code: simultaneous legacy and modern relations should not be allowed.

File: tests/integration/relations/test_relations.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                num_units=1,
+                series=CHARM_SERIES,
+                config={"profile": "testing"},
+            ),
+            ops_test.model.deploy(
+                MAILMAN3_CORE_APP_NAME,
+                application_name=MAILMAN3_CORE_APP_NAME,
+                channel="stable",
+                config={"hostname": "example.org"},
+            ),
```
Comment (marceloneppel) [inline tests/integration/relations/test_relations.py:53]: We need to pass `series="focal"` here to support Juju 3.4.

File: src/relations/postgresql_provider.py [hunk 2]
```diff
@@ -199,6 +215,36 @@ def _update_unit_status(self, relation: Relation) -> None:
...
+
+        # Emit a database requested event if the setup key (database name and optional
+        # extra user roles) was added to the relation databag by the application.
+        if "database" in _diff.added:
+            getattr(self.database_provides.on, "database_requested").emit(
+                event.relation, app=event.app, unit=event.unit
+            )
+
+    def _update_unit_status_on_blocking_endpoint_simultaneously(self):
+        """# Clean up Blocked status if this is due related of multiple endpoints."""
```
Comment (marceloneppel) [inline src/relations/postgresql_provider.py:240]:
```suggestion
        """Clean up Blocked status if this is due related of multiple endpoints."""
```

File: src/relations/postgresql_provider.py [hunk 3]
```diff
@@ -190,6 +199,13 @@ def update_endpoints(self, event: DatabaseRequestedEvent = None) -> None:
                 read_only_endpoints,
             )
 
+    def _check_multiple_endpoints(self) -> bool:
+        """Checks if there are relations with other endpoints."""
+        relation_names = [relation.name for relation in self.charm.client_relations]
+        if "database" in relation_names and len(relation_names) > 1:
+            return True
```
Comment (marceloneppel) [inline src/relations/postgresql_provider.py:206]: Can this return `True` if there are multiple applications related to the `database` endpoint?

File: src/relations/db.py [hunk 4]
```diff
@@ -278,6 +300,15 @@ def _update_unit_status(self, relation: Relation) -> None:
             if not self._check_for_blocking_relations(relation.id):
                 self.charm.unit.status = ActiveStatus()
 
+    def _update_unit_status_on_blocking_endpoint_simultaneously(self):
```
Comment (marceloneppel) [inline src/relations/db.py:303]: Are you calling this method?

**Review by marceloneppel — COMMENTED** [review]
Thanks, BalabaDmintri! That's a great implementation. I left some comments to improve the code a bit.

File: tests/integration/relations/test_relations.py (hunk 1 above)
Comment (BalabaDmintri) [reply tests/integration/relations/test_relations.py:53]: @marceloneppel this is only used for mailman3-core?

**Comment by taurus-forever** [comment]
@BalabaDmintri can you please update PR description. It doesn't represent the content. Tnx!

File: src/relations/postgresql_provider.py (hunk 3 above)
Comment (marceloneppel) [reply src/relations/postgresql_provider.py:206]: I think the logic from `src/relations/db.py`'s `_check_multiple_endpoints` will help you here.

File: src/relations/postgresql_provider.py [hunk 5]
```diff
@@ -190,6 +199,13 @@ def update_endpoints(self, event: DatabaseRequestedEvent = None) -> None:
                 read_only_endpoints,
             )
 
+    def _check_multiple_endpoints(self) -> bool:
+        """Checks if there are relations with other endpoints."""
+        relation_names = {relation.name for relation in self.charm.client_relations}
```
Comment (BalabaDmintri) [inline src/relations/postgresql_provider.py:204]: @marceloneppel I fixed it, like `set`
or use the implementation from `src/relations/db.py's` `_check_multiple_endpoints` ?

File: tests/integration/relations/__init__.py [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# Copyright 2023 Canonical Ltd.
```
Comment (taurus-forever) [inline tests/integration/relations/__init__.py:1]: nit: 2024, you can ignore it here and update the next PR.

File: tests/integration/relations/test_relations.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                charm,
+                application_name=APP_NAME,
+                num_units=1,
+                series=CHARM_SERIES,
+                config={"profile": "testing"},
+            ),
+            ops_test.model.deploy(
+                MAILMAN3_CORE_APP_NAME,
+                application_name=MAILMAN3_CORE_APP_NAME,
+                channel="stable",
```
Comment (taurus-forever) [inline tests/integration/relations/test_relations.py:51]: This charm has no updates for ages... and I would recommend to use 2 postgresql-test-app instead... but let's see how stable the current approach is.

**Review by taurus-forever — APPROVED** [review]
LGTM, @dragomirp consider to merge if LGTM as well. Tnx!

**Comment by dragomirp** [comment]
Hi, @BalabaDmintri, please fix the error in the unit test and sync up with main.

**Comment by BalabaDmitri** [comment]
> Hi, @BalabaDmintri, please fix the error in the unit test and sync up with main.

**Comment by BalabaDmitri** [comment]
@dragomirp Can  you check

**Comment by BalabaDmitri** [comment]
@dragomirp Can you run actions

**Review by dragomirp — APPROVED** [review]

**Review by marceloneppel — APPROVED** [review]
